### PR TITLE
Retain CoreDNS corefile when migration fails in kubeadm

### LINKED ISSUE #84326: Upgrade to 1.16.x failed if CoreDNS was previously deployed with self-built image
**What happened**:
Upgrade using kubeadm failed because of CoreDNS migration enchancement

**What you expected to happen**:
Upgrade passed correctly

**How to reproduce it (as minimally and precisely as possible)**:
1. Use any 1.15.x Kubernetes with CoreDNS that deployed with any unusual image (self built for example). We using commits counter as part of image tag, e.g - 1.6.2-100. 
2. Try to upgrade to 1.16.x with kubeadm.

**Anything else we need to know?**:
Logs part without ignoring preflight:
```
[preflight] Some fatal errors occurred:
    [ERROR CoreDNSUnsupportedPlugins]: start version 'v1.6.2-100' not supported
    [ERROR CoreDNSMigration]: start version 'v1.6.2-100' not supported
```
Logs part with ignoring preflight:
```
unable to migrate CoreDNS ConfigMap: start version 'v1.6.2-100' not supported
[upgrade/postupgrade] FATAL post-upgrade error
k8s.io/kubernetes/cmd/kubeadm/app/cmd/upgrade.runApply
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/upgrade/apply.go:182
k8s.io/kubernetes/cmd/kubeadm/app/cmd/upgrade.NewCmdApply.func1
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/upgrade/apply.go:80
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:830
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:914
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:864
k8s.io/kubernetes/cmd/kubeadm/app.Run
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/kubeadm.go:50
main.main
    _output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/kubeadm.go:25
runtime.main
    /usr/local/go/src/runtime/proc.go:200
runtime.goexit
    /usr/local/go/src/runtime/asm_amd64.s:1337
```

**Environment**:
- Kubernetes version (use `kubectl version`):
```
root@kaas-node-20262e33-f6a9-11e9-a537-02420a0f0f01:~# kubectl version
Client Version: version.Info{Major:"1", Minor:"15+", GitVersion:"v1.15.3-8+d06f4e3032941e", GitCommit:"d06f4e3032941e9e8099845fdde660d482f571fb", GitTreeState:"clean", BuildDate:"2019-09-13T14:22:14Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"15+", GitVersion:"v1.15.3-8+d06f4e3032941e", GitCommit:"d06f4e3032941e9e8099845fdde660d482f571fb", GitTreeState:"clean", BuildDate:"2019-09-13T14:24:40Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
OpenStack
- OS (e.g: `cat /etc/os-release`):
```
NAME="Ubuntu"
VERSION="18.04.2 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.2 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```
- Kernel (e.g. `uname -a`):
`Linux kube-node-01 4.15.0-51-generic #55-Ubuntu SMP Wed May 15 14:27:21 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools:
Kubeadm

### PR DESCRIPTION
**Pull request by rajansandeep**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Currently, when a user chooses to skip the preflight check of CoreDNS corefile migration, the kubeadm upgrade fails. This fix bypasses the migration step and retains the existing Corefile in case there is a preflight error and the user chooses to skip.

**Which issue(s) this PR fixes**:

Fixes #84326

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: cmd/kubeadm/app/phases/addons/dns/dns.go [hunk 1]
```diff
@@ -238,7 +238,12 @@ func createCoreDNSAddon(deploymentBytes, serviceBytes, configBytes []byte, clien
 	}
 	if IsCoreDNSConfigMapMigrationRequired(corefile) {
 		if err := migrateCoreDNSCorefile(client, coreDNSConfigMap, corefile, currentInstalledCoreDNSVersion); err != nil {
-			return err
+			// Errors in Corefile Migration is verified during preflight checks. This part will be executed should a user choose
+			// to ignore preflight check errors.
+			klog.V(2).Infof("the Corefile Migration did not occur due to an error: %v. The existing CoreDNS Corefile configuration has been retained.", err)
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]: probably better to have this as a `klog.Warning(`

File: cmd/kubeadm/app/phases/upgrade/preflight.go [hunk 2]
```diff
@@ -108,7 +108,7 @@ func checkMigration(client clientset.Interface) error {
 
 	_, err = migration.Migrate(currentInstalledCoreDNSversion, kubeadmconstants.CoreDNSVersion, corefile, false)
 	if err != nil {
-		return err
+		return errors.Errorf("the CoreDNS configuration migration will not occur due to an error: %v", err)
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/upgrade/preflight.go]: should be `errors.Wrap(err, ....)`

File: cmd/kubeadm/app/phases/addons/dns/dns.go (hunk 1 above)
Comment (neolit123) [reply cmd/kubeadm/app/phases/addons/dns/dns.go]: `a user choose` -> `a user chose`
should `chose` be past tense here?

File: cmd/kubeadm/app/phases/upgrade/preflight.go (hunk 2 above)
Comment (chrisohaver) [inline cmd/kubeadm/app/phases/upgrade/preflight.go]:
```suggestion
		return errors.Errorf("the CoreDNS configuration will not be migrated, and may be incompatible with the upgraded version of CoreDNS: %v", err)
```

File: cmd/kubeadm/app/phases/addons/dns/dns.go (hunk 1 above)
Comment (chrisohaver) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]:
```suggestion
			klog.V(2).Infof("the CoreDNS Configuration was not migrated: %v. The existing CoreDNS Corefile configuration has been retained.", err)
```

**Comment by neolit123** [comment]
thanks @rajansandeep 
just noting, i think this is the first time we add logic in kubeadm where ignoring a preflight check propagates into areas outside of preflight. but it feels fine to me.

/priority important-longterm
/approve
leaving LGTM to the folks on CC.

**Comment by neolit123** [comment]
/hold
for the minor nits.

**Review by chrisohaver — COMMENTED** [review]
wording tweaks, and spell out possible repercussions of not migrating the coredns config

**Comment by chrisohaver** [comment]
IMO, this fix isn't going to solve custom CoreDNS problems alone.

It needs to be documented somewhere that users with custom CoreDNS need to use upgrade phases instead, skipping the CoreDNS upgrade, and doing that upgrade manually (and manually migrating CoreDNS config if necessary).
If a user of a custom CoreDNS upgrades using the normal kubeadm upgrade process, and ignores the warning, then their custom CoreDNS will get replaced with a stock version of CoreDNS.  In most cases this would result in a non-functioning DNS after the upgrade finishes. The most common reason for using a custom CoreDNS compile is to use external plugins, and an external plugin present in a Corefile will cause a stock CoreDNS to exit before starting.

Perhaps we could succinctly explain this in the preflight warning...

"... clusters with CoreDNS deployments should use upgrade phases to skip CoreDNS upgrade ..."

**Comment by neolit123** [comment]
> "... clusters with CoreDNS deployments should use upgrade phases to skip CoreDNS upgrade ..."

adding something like that SGTM, we can do that once phase support for `upgrade apply` is added:
https://github.com/kubernetes/kubeadm/issues/1318

**Comment by neolit123** [comment]
> I don't think this change actually does that.

my mistake, it does not.

**Comment by chrisohaver** [comment]
Actually, now i'm thinking it makes more sense retain both the original corefile AND the original coredns deployment when the migration fails for any reason.

That changes the preflight warning to something along the lines of: "CoreDNS will not be upgraded: (error)"

**Comment by neolit123** [comment]
@rajansandeep WDYT about @chrisohaver 's last comment?
it seems fine to me.

**Comment by rajansandeep** [comment]
>Actually, now i'm thinking it makes more sense retain both the original corefile AND the original coredns deployment when the migration fails for any reason.
That changes the preflight warning to something along the lines of: "CoreDNS will not be upgraded: (error)"

@neolit123 Yes, I'll push out a commit to make the changes @chrisohaver suggested.

File: cmd/kubeadm/app/phases/upgrade/preflight.go [hunk 3]
```diff
@@ -108,7 +108,7 @@ func checkMigration(client clientset.Interface) error {
 
 	_, err = migration.Migrate(currentInstalledCoreDNSversion, kubeadmconstants.CoreDNSVersion, corefile, false)
 	if err != nil {
-		return errors.Wrap(err, "the CoreDNS configuration will not be migrated, and may be incompatible with the upgraded version of CoreDNS")
+		return errors.Wrap(err, "CoreDNS will not be upgraded")
```
Comment (chrisohaver) [inline cmd/kubeadm/app/phases/upgrade/preflight.go]: I think we should include the error in the message.

**Comment by neolit123** [comment]
SGTM, thanks for the updates.
the PR is approved.
leaving LGTM to @chrisohaver

**Pull request merged**